packages/protocol/contracts/Median.sol

```diff
@@ -0,0 +1,441 @@
+   * even-length array. It's more gas efficient than calling `compute` twice. 
+   * @param arr An array of signed integers.
+   * @param k
+   * @return a tuple containing the kth and (k+1)st smallest elements in `arr`,
+   * respectively
+   */ 
+  function compute2
+  (
+    int256[] memory arr,
+    uint256 k
+  )
+    external
+    pure
+    returns (int256, int256)
+  {
+    return compute2InPlace(copy(arr), k);
+  }
+
+  /* Returns the kth smallest element in an array of signed ints.
+   * @dev The input array `arr` may be modified during the computation.
+   * @param arr An array of unsigned integers.
+   * @param k
+   * @param inplace If true, original array is not modified. Otherwise, the
+   *        array may be modified.
+   * @return the kth smallest elements in `arr`
+   */ 
+  function computeInPlace
+  (
+    int256[] memory arr,
+    uint256 k
+  )
+    public
+    pure
+    returns (int256)
+  {
+    assert(0 <= k && k <= arr.length - 1);
+    if (arr.length <= SMALL_ARRAY_MAX_LENGTH) {
+      return _selectK_smallArray(arr, k);
+    }
+    (uint256 idx1, uint256 idx2) = _quickSelect(arr, 0, arr.length - 1, k, false);
+    return arr[idx1];
+  }
+
+  /* Returns the kth and (k+1)st smallest elements in an array of signed ints without
+   * modifying the input array.
+   * @dev This is for when one wants to compute the exact median of an
+   * even-length array. It's more gas efficient than calling `compute` twice.
+   * Note that array `arr` may be modified during the computation. 
+   * @param arr An array of signed integers.
+   * @param k
+   * @return a tuple containing the kth and (k+1)st smallest elements in `arr`,
+   * respectively
+   */ 
+  function compute2InPlace
+  (
+    int256[] memory arr,
+    uint256 k
+  )
+    public
+    pure
+    returns (int256, int256)
+  {
+    assert(0 <= k && k <= arr.length - 2);
+    if (arr.length <= SMALL_ARRAY_MAX_LENGTH) {
+      int256 x1 = _selectK_smallArray(arr, k);
+      int256 x2 = arr[k + 1];
+      return (x1, x2);
+    }
+    (uint256 idx1, uint256 idx2) = _quickSelect(arr, 0, arr.length - 1, k, true);
+    return (arr[idx1], arr[idx2]);
+  }
+
+  // PRIVATE FUNCTIONS
+
+  /* Select the index of the kth element in an array.
+   * @dev This function may modify array.
+   * @param arr an array of uints.
+   * @param left the left index to begin search.
+   * @param right the right index to begin search.
+   * @param k
+   * @param selectKplusOne a bool representing whether the function should
+   *                       return the (k+1)st element or not.
+   * @return a tuple (i, j) where i and j are the indices of the kth and
+   *         (k+1)st elements of `arr`, respectively. In the case where
+   *         `selectKplusOne` is false, the tuple returned is (i, 0).
+   */
+   function _quickSelect
+  (
+    int256[] memory arr,
+    uint256 left,
+    uint256 right,
+    uint256 k,
+    bool selectKplusOne
```

Comment:
It's weird how these are sometimes implemented as different functions (`compute`/`computeInPlace`) and sometimes parameterized by a flag

Reply:
`compute`/`computeInPlace` are public functions. `quickselect` is a private function so it's optimized for gas cost savings without caring how one interfaces with it. So there's just one function that handles many cases.

Reply:
You said smth about not wanting users to pass too many arguments/flags to the main/external functions.

Reply:
Got it. I think these flags are bad in general, not only for external functions. Solidity is strongly typed, you're working around that. When `selectKplusOne` is `true`, both the returned type becomes a tuple and what the function does changes so `quickSelect` is not a valid name anymore. The two should be implemented as separate functions.